### PR TITLE
Add support for "#rrggbbaa" colors

### PR DESCRIPTION
**Pull request by cookieguru**
As requested in #1142.

I forsee a new chapter in [Oh shit, git!](http://www.ohshitgit.com/).

### CODE REVIEW

File: doc/types.md [hunk 1]
```diff
@@ -51,8 +51,10 @@ The new value of the changed property.
 
 Colors are specified as strings using one of the following formats:
 
-* `#xxxxxx`
-* `#xxx`
+* `#xxxxxx` Hexidecimal rgb
```
Comment (ralfstx) [inline doc/types.md]: typo

File: doc/types.md [hunk 2]
```diff
@@ -51,8 +51,10 @@ The new value of the changed property.
...
+* `#xxx` Hexidecimal rgb
```
Comment (ralfstx) [inline doc/types.md]: typo

File: src/tabris/util-colors.js [hunk 3]
```diff
@@ -29,6 +29,24 @@ export function colorStringToArray(str) {
       255
     ];
   }
+  // #xxxxxxxx
+  if (/^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/.test(str)) {
```
Comment (ralfstx) [inline src/tabris/util-colors.js]: Couldn't we save code and cycles by making the last position optional and so combine the regexp for 8 and 6 hex digit colors (and also for 4 and 3 digits)?
```js
/^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})?$/
```
For 6-digit hex numbers, `RegExp.$4` would then be the empty string.

**Comment by cookieguru** [comment]
@ralfstx Changes added

**Comment by ralfstx** [comment]
@cookieguru Looks good! Why didn't you do the same for 3 and 4 digit hex strings?

**Comment by cookieguru** [comment]
@ralfstx Didn't think of it.  But it's in there now.

On the note of combining, the rgb(a) syntaxes could be combined:
```js
/^rgba?\(\s*(0|[1-9]\d?|1\d\d?|2[0-4]\d|25[0-5])\s*,\s*(0|[1-9]\d?|1\d\d?|2[0-4]\d|25[0-5])\s*,\s*(0|[1-9]\d?|1\d\d?|2[0-4]\d|25[0-5])\s*(?:,\s*((0.[1-9])|[01])\s*)?\)$/
```
Only problem is this test fails:
```js
it('rejects rgba function strings with wrong argument count', function() {
  expect(() => {
    colorStringToArray('rgba(23, 42, 47)');
  }).to.throw();
});
```

File: src/tabris/util-colors.js [hunk 4]
```diff
@@ -11,22 +11,31 @@ export function colorStringToArray(str) {
   if (str === 'transparent') {
     return [0, 0, 0, 0];
   }
-  // #xxxxxx
-  if (/^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/.test(str)) {
+  // #xxxxxx and #xxxxxxxx
+  if (/^#([0-9A-F]{2})([0-9A-F]{2})([0-9A-F]{2})([0-9A-F]{2})?$/i.test(str)) {
```
Comment (ralfstx) [inline src/tabris/util-colors.js]: `a-fA-F` was used on purpose as case insensitive regex are known to be slower.
http://stackoverflow.com/questions/32010/is-regex-case-insensitivity-slower
http://stackoverflow.com/questions/40353587/a-za-z-vs-a-z-i-case-insensitive

File: src/tabris/util-colors.js [hunk 5]
```diff
@@ -11,22 +11,31 @@ export function colorStringToArray(str) {
...
+  if (/^#([0-9A-F])([0-9A-F])([0-9A-F])([0-9A-F])?$/i.test(str)) {
     return [
       parseInt(RegExp.$1, 16) * 17,
       parseInt(RegExp.$2, 16) * 17,
       parseInt(RegExp.$3, 16) * 17,
-      255
+      RegExp.$4 === '' ? 255 : parseInt(RegExp.$4, 16) * 17
+    ];
+  }
+  // #xxxx
```
Comment (ralfstx) [inline src/tabris/util-colors.js]: This block should now be removed.

**Comment by ralfstx** [comment]
@cookieguru Now you forgot to remove the block for 4-digit strings.

The regex for rgb cannot be easily combined as the function names differ. We don't want to allow `rgb()` with 4 arguments or `rgba()` with only 3. However, let's focus on this change now. It's pending for too long already.

**Comment by cookieguru** [comment]
@ralfstx Interesting info on case sensitivity!  I've updated the PR.

**Comment by ralfstx** [comment]
Thanks for your patience, @cookieguru!

**Comment by cookieguru** [comment]
Thanks for your guidance @ralfstx!

**Pull request merged**